### PR TITLE
[occm] Tls exclude ports

### PR DESCRIPTION
**Pull request by reneluria**
[oocm] Add annotation to exclude ports from TERMINATED_HTTPS: tls-ports-exclude

**What this PR does / why we need it**:

It adds support for the annotation `loadbalancer.openstack.org/service-ssl-ports`.
When using `loadbalancer.openstack.org/default-tls-container-ref`, all listeners for ports included in this annotations are creating using TERMINATED_HTTPS

For a multi-port service, having for example, ports 80 and 443, we want some ports to be excluded and use proto HTTP instead, to have plain http on some ports and https on the others

Example:

```yaml
apiVersion: v1
kind: Service
metadata:
  labels:
    app: myapp
  name: myapp
  annotations:
    loadbalancer.openstack.org/x-forwarded-for: "true"
    loadbalancer.openstack.org/default-tls-container-ref: "https://api.pub1.infomaniak.cloud/keymanager/v1/containers/e53fd5d0-c3d0-4707-8c20-bbf11796e1b5"
    loadbalancer.openstack.org/service-ssl-ports: "443"
spec:
  ports:
  - port: 80
    protocol: TCP
    targetPort: 80
  - port: 443
    protocol: TCP
    targetPort: 443
  selector:
    app: myapp
  type: LoadBalancer
status:
  loadBalancer: {}
```

**Which issue this PR fixes(if applicable)**:
-none-

**Special notes for reviewers**:

Create the above service with this image, listing listeners for this load balancer via the openstack api, you should get:

```
+--------------------------------------+--------------------------------------+------+----------------------------------+------------------+---------------+----------------+
| id                                   | default_pool_id                      | name | project_id                       | protocol         | protocol_port | admin_state_up |
+--------------------------------------+--------------------------------------+------+----------------------------------+------------------+---------------+----------------+
| 7f8295ee-ae54-4025-8c12-1ce0695f45b8 | 9ff787d0-87fe-4c2a-ba74-6a819fdf006c |      | 4c08a353d2414273aa43ee0bf6ec2683 | HTTP             |            80 | True           |
| b1e17721-83c8-44d3-ae5a-63b4bd7fcefe | fbeedf15-36d0-45b9-a181-be47d66430cf |      | 4c08a353d2414273aa43ee0bf6ec2683 | TERMINATED_HTTPS |           443 | True           |
+--------------------------------------+--------------------------------------+------+----------------------------------+------------------+---------------+----------------+
```

**Release note**:
```release-note
NONE
```

### CODE REVIEW

**Comment by lingxiankong** [comment]
Rather than `loadbalancer.openstack.org/tls-ports-exclude: "80"`, I'd prefer using something like `loadbalancer.openstack.org/service-ssl-ports: "443,8443"` to be more straightforward. Thoughts?

**Comment by reneluria** [comment]
> Rather than `loadbalancer.openstack.org/tls-ports-exclude: "80"`, I'd prefer using something like `loadbalancer.openstack.org/service-ssl-ports: "443,8443"` to be more straightforward. Thoughts?

yes, why not, here is the new version

**Review by lingxiankong — CHANGES_REQUESTED** [review]
Does this new annotation support service update?

File: pkg/openstack/loadbalancer.go [hunk 1]
```diff
@@ -1491,7 +1493,13 @@ func (lbaas *LbaasV2) ensureOctaviaListener(lbID string, oldListeners []listener
 
 	proto := toListenersProtocol(port.Protocol)
 	if svcConf.tlsContainerRef != "" {
-		proto = listeners.ProtocolTerminatedHTTPS
+		// check if we need to exclude port from tls
+		if tlsIsPortIncluded(svcConf.tlsServicePorts, port.Port) {
+			proto = listeners.ProtocolTerminatedHTTPS
+		} else {
+			klog.V(2).Infof("No tlsContainer for excluded port %v", port.Port)
```
[the remark this answers is not in the document]
Comment (reneluria) [reply pkg/openstack/loadbalancer.go:1500]: that print statement is not needed, just removed it

File: pkg/openstack/loadbalancer.go [hunk 2]
```diff
@@ -3393,3 +3403,35 @@ func PreserveGopherError(rawError error) error {
...
+func GetTlsServicePorts(service *corev1.Service) ([]int32) {
+	var noPorts []int32
+	if annotationValue, ok := service.Annotations[ServiceAnnotationTlsServicePorts]; ok {
+		portStrArray := strings.Split(annotationValue, ",")
+		ports := make([]int32, len(portStrArray))
+		for i, value := range(portStrArray) {
+			port, err := strconv.Atoi(value)
+			if err != nil {
+				klog.Warningf("could not parse %v into int: %v", value, err)
+				return noPorts
```
[the remark this answers is not in the document]
Comment (reneluria) [reply pkg/openstack/loadbalancer.go]: good point, fixed

File: pkg/openstack/loadbalancer.go [hunk 3]
```diff
@@ -3393,3 +3403,35 @@ func PreserveGopherError(rawError error) error {
 	}
 	return rawError
 }
+
+// GetTlsServicePorts gets a list of ports from the ServiceAnnotationTlsPortsExclude
+// annotation
+// the format is a comma separated list of numeric ports
+func GetTlsServicePorts(service *corev1.Service) ([]int32) {
```
[the remark this answers is not in the document]
Comment (reneluria) [reply pkg/openstack/loadbalancer.go]: good point, removed

File: pkg/openstack/loadbalancer.go [hunk 4]
```diff
@@ -3393,3 +3403,35 @@ func PreserveGopherError(rawError error) error {
...
+			}
+			ports[i] = int32(port)
+		}
+		return ports
+	}
+	return noPorts
+}
+
+// Tests if a port is in the list of excluded ports
+func tlsIsPortIncluded(includedPorts []int32, port int32) bool {
```
[the remark this answers is not in the document]
Comment (reneluria) [reply pkg/openstack/loadbalancer.go]: here you go

**Pull request closed without merging**